doc/source/development/rfc/rfc86_column_oriented_api.rst

```diff
@@ -0,0 +1,692 @@
+record batches being an ArrowArray of a subset of features. Instead of iterating
+over individual features, one iterates over a batch of several features at
+once.
+
+The ArrowArrayStream, ArrowSchema, ArrowArray structures are defined in a ogr_recordbatch.h
+public header file, directly derived from https://github.com/apache/arrow/blob/master/cpp/src/arrow/c/abi.h
+to get API/ABI compatibility with Apache Arrow C++. This header file must be
+explicitly included when the related array batch API is used.
+
+The following virtual method is added to the OGRLayer class:
+
+  .. code-block:: cpp
+
+        virtual bool OGRLayer::GetArrowStream(struct ArrowArrayStream* out_stream,
+                                              CSLConstList papszOptions = nullptr);
+
+This method is also available in the C API as OGR_L_GetArrowStream().
+
+out_stream is a pointer to a ArrowArrayStream structure, that can be in a uninitialized
+state (the method will ignore any initial content).
+
+On successful return, and when the stream interfaces is no longer needed, it must must
+be freed with out_stream->release(out_stream).
+
+There are extra precautions to take into account in a OGR context. Unless
+otherwise specified by a particular driver implementation, the ArrowArrayStream
+structure, and the ArrowSchema or ArrowArray objects its callbacks have returned,
+should no longer be used (except for potentially being released) after the
+OGRLayer from which it was initialized has been destroyed (typically at dataset
+closing). Furthermore, unless otherwise specified by a particular driver
+implementation, only one ArrowArrayStream can be active at a time on
+a given layer (that is the last active one must be explicitly released before
+a next one is asked). Changing filter state, ignored columns, modifying the schema
+or using ResetReading()/GetNextFeature() while using a ArrowArrayStream is
+strongly discouraged and may lead to unexpected results. As a rule of thumb,
+no OGRLayer methods that affect the state of a layer should be called on a
+layer, while an ArrowArrayStream on it is active.
+
+A potential usage can be:
+
+.. code-block:: cpp
+
+    struct ArrowArrayStream stream;
+    if( !poLayer->GetArrowStream(&stream, nullptr))
+    {
+        fprintf(stderr, "GetArrowStream() failed\n");
+        exit(1);
+    }
+    struct ArrowSchema schema;
+    if( stream.get_schema(&stream, &schema) == 0 )
+    {
+        // Do something useful
+        schema.release(schema);
+    }
+    while( true )
+    {
+        struct ArrowArray array;
+        // Look for an error (get_next() returning a non-zero code), or
+        // end of iteration (array.release == nullptr)
+        //
+        if( stream.get_next(&stream, &array) != 0 ||
+            array.release == nullptr )
+        {
+            break;
+        }
+        // Do something useful
+        array.release(&array);
+    }
+    stream.release(&stream);
```

Comment:
Why does this method require the explicit object pointer argument?

Reply:
The release callbacks may need to use the private_data member of the ArrowArray and ArrowArrayStream structures (those are C structures, not C++ objects)

Reply:
Hmm, ok. Do you think it's worth having a gdal specific wrapper around those structures so that we can provide a nicer public api which hides this kind of detail?

Reply:
> Do you think it's worth having a gdal specific wrapper around those structures so that we can provide a nicer public api which hides this kind of detail?

The Arrow C stream/array interfaces defines a standard, and I'd want to avoid defining something GDAL specific on top of it. Anything that would make it more friendly to consume it would apply to any software that would produce Arrow C streams/arrays, so that doesn't really belong to GDAL specificially.  @paleolimbot  raised a discussion about that in https://lists.apache.org/thread/6vt7btpbrkm554otxqz7oxzdlpfwkhwc . He has prototyped
https://github.com/paleolimbot/gpkg/tree/master/src/arrow-hpp for the writing side. Not sure if there's an equivalent on the reading side (except using the arrow-cpp library itself to ingest Arrow C arrays and expose corresponding C++ objects, but that's an heavy dependency)